doc/charm++/manual.rst

```diff
@@ -2894,96 +2922,85 @@ elements migratable (see migration section above) and choose a load
 description of available load balancing strategies).
 
-There are three different ways to use load balancing for chare arrays to
-meet different needs of the applications. These methods are different in
-how and when a load balancing phase starts. The three methods are:
-**periodic load balancing mode**, **at sync mode** and **manual mode**.
-
-In *periodic load balancing mode*, a user specifies only how often load
-balancing is to occur, using +LBPeriod runtime option to specify the
-time interval.
-
-In *at sync mode*, the application invokes the load balancer explicitly
-at appropriate (generally at a pre-existing synchronization boundary) to
-trigger load balancing by inserting a function call (AtSync) in the
-application source code.
-
-In the prior two load balancing modes, users do not need to worry about
-how to start load balancing. However, in one scenario, those automatic
-load balancers will fail to work - when array elements are created by
-dynamic insertion. This is because the above two load balancing modes
-require an application to have fixed the number of objects at the time
-of load balancing. The array manager needs to maintain a head count of
-local array elements for the local barrier. In this case, the
-application must use the *manual mode* to trigger load balancer.
+There are three different ways to use load balancing for chare arrays
+to meet different needs of the applications. These methods are
+different in how and when a load balancing phase starts. The two
+methods are: **AtSync mode** and **periodic load balancing mode**.
+
+In *periodic load balancing mode*, a user specifies only how often
+load balancing is to occur, using the *+LBPeriod* runtime option to
+specify the time interval.
+
+In *AtSync mode*, the application invokes the load balancer explicitly
+at an appropriate location (generally at a pre-existing
+synchronization boundary) to trigger load balancing by inserting a
+function call (``AtSync()``) in the application source code.
+
+Note that applications that use dynamically insertion or deletion of
+array elements must not be doing so when any element calls
+``AtSync()``. This is because AtSync mode requires an application to
+have a fixed, known number of objects when determining if ``AtSync()``
+has been called by all relevant objects in order to prevent race
+conditions (the implementation is designed to be robust against these
+issues and will often be able to handle them, but we make no
+guarantees if these rules are not obeyed). If using dynamic insertion,
+please ensure that insertions and calls to ``AtSync()`` cannot be
+interleaved and that ``doneInserting()`` is called after insertions
+are complete and before any element calls ``AtSync()``. Insertions
+and/or deletions may begin again after load balancing is complete
+(i.e. ``ResumeFromSync()`` is called for an object on the given PE for
+insertions or for the object in question for deletions).
```

Comment:
Yes, exactly that. The criterion for deletions is actually stricter than the truth, really you should be in the clear so long as _any_ of the object on the PE of object to be deleted have had a `ResumeFromSync()` call. The trouble, of course, is that you don't know what PE an object is on when coming out of an LB invocation until that _exact_ object has been resumed, as it may have migrated during load balancing.